wgpu-core/src/device/life.rs

```diff
@@ -674,28 +676,34 @@ impl<A: HalApi> LifetimeTracker<A> {
                     }
                     _ => panic!("No pending mapping."),
                 };
-                let status = if mapping.range.start != mapping.range.end {
+
+                let size = mapping.range.end - mapping.range.start;
+                let host = mapping.op.host;
+                let ptr_maybe = if size != 0 {
                     log::debug!("Buffer {:?} map state -> Active", buffer_id);
-                    let host = mapping.op.host;
-                    let size = mapping.range.end - mapping.range.start;
                     match super::map_buffer(raw, buffer, mapping.range.start, size, host) {
-                        Ok(ptr) => {
-                            buffer.map_state = resource::BufferMapState::Active {
-                                ptr,
-                                range: mapping.range.start..mapping.range.start + size,
-                                host,
-                            };
-                            resource::BufferMapAsyncStatus::Success
-                        }
+                        Ok(ptr) => Some(ptr),
                         Err(e) => {
                             log::error!("Mapping failed {:?}", e);
-                            resource::BufferMapAsyncStatus::Error
+                            None
                         }
                     }
                 } else {
-                    resource::BufferMapAsyncStatus::Success
+                    ptr::NonNull::new(self.empty_mapping.as_mut_ptr())
```

Comment:
Couldn't we just do `ptr::NonNull::new((&mut []).as_mut_ptr())`? We'd be recreating a zero sized slice on the other side, unless for some reason gecko needs to deref this pointer.

Reply:
I don't think this is valid. A `*mut u8` isn't holding `[]` alive, so the `[]` will die instantly, and the pointer becomes invalid.

Reply:
That's the trick though, a zero size slice is pointing to nothing, which is why `&mut []` is of lifetime `'static`. The way that you can show this is that it's valid to create a zero-size slice to an empty vector which hasn't allocated on the heap at all.

Reply:
that's tricky indeed :/